### ISSUE TITLE
Display key in dynamic task mapping using dict in UI view of Mapped Tasks' map index

### ISSUE THREAD

**Issue by kiaradlf** [apache/airflow#39390]
### Apache Airflow version

2.9.0

### What happened?

As per the confluence wiki, dynamic task mapping [users will have an option to click on any generated task (e.g. run_after_loop[0], run_after_loop[1] **or run_after_loop[key] if the map function is iterating over a dictionary**)](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-42+Dynamic+Task+Mapping) (emphasis mine).
However, if I use dynamic task mapping with `.expand()` over a dictionary, the UI view Mapped Tasks still shows a numerical Map Index for the sub-tasks.

### What you think should happen instead?

It would be nice if it might instead pick up on keys from such dictionary inputs such that one would not have to manually specify `map_index_template` (alongside setting such a key in the context).

### How to reproduce

- `.expand()` over a dictionary
- open an operator's web UI view Mapped Tasks
- find a numerical Map Index for the sub-tasks

### Operating System

ChromeOS

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by fredthomsen** [comment]
Came across a use-case I had this week that this would solve.  If this is a desired feature, go ahead and assign it to me and I'll submit a PR.

**Comment by uranusjr** [comment]
Using `map_index_template` would be a solution. I don’t think it currently receives the dict you used to expand against, but you can pass in the dict by other means.

**Comment by fredthomsen** [comment]
Thanks @uranusjr .  That is the path I ended up taking.  I initially thought there might be a use-case here that `map_index_template` did not meet; however, that is not the case.

**Comment by kiaradlf** [comment]
won't `map_index_template` only be reflected after a run completes?

**Comment by fredthomsen** [comment]
@kiaradlf yes that is correct.  If that is your use case then `map_index_template` will not help you.
